File: backends/spring-boot/src/main/java/com/hellotimepro/springboot/web/HealthController.java
package com.hellotimepro.springboot.web;

import com.hellotimepro.springboot.config.AppProperties;
import com.hellotimepro.springboot.dto.Dtos.Avatar;
import com.hellotimepro.springboot.dto.Dtos.Envelope;
import com.hellotimepro.springboot.dto.Dtos.HealthData;
import com.hellotimepro.springboot.dto.Dtos.StackInfo;
import com.hellotimepro.springboot.dto.Dtos.StackItem;
import com.hellotimepro.springboot.service.AvatarService;
import java.lang.management.ManagementFactory;
import java.util.List;
import org.springframework.boot.SpringBootVersion;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1")
public class HealthController {
  private final AppProperties props;
  private final AvatarService avatars;

  public HealthController(AppProperties props, AvatarService avatars) {
    this.props = props;
    this.avatars = avatars;
  }

  @GetMapping("/health")
  public Envelope<HealthData> health() {
    boolean isSqlite = props.getDbDriver().equals("sqlite");
    String dbName = isSqlite ? "SQLite" : "PostgreSQL";
    String dbIcon = isSqlite ? "sqlite" : "postgresql";
    String dbVer  = isSqlite ? "3" : "16";

    String summary =
        "基于 Java + Spring Boot 3 + Spring Data JPA 核心骨架，" +
        "选用 Hibernate 作为 JPA 提供商，仓库级 scripts/db 维护数据库 schema，" +
        "Jackson 负责 JSON 序列化，Jakarta Bean Validation 实现声明式数据校验，" +
        "同时支持 PostgreSQL 与 SQLite 双数据库驱动切换。" +
        "利用 Spring 核心的控制反转与依赖注入容器，以单例模式管理组件，" +
        "Spring Boot 自动配置开箱即用。" +
        "Spring Data JPA 通过方法名自动解析生成 SQL 查询，" +
        "针对跨库时间列差异自制了全局属性转换器，并配置了 JPA 级别悲观锁确保行级并发一致性。" +
        "通过在 DTO 记录类属性上贴附 Bean Validation 注解完成请求边界拦截，结合注解声明式路由映射。" +
        "基于 Spring AOP 机制使用 @Transactional 实现无侵入的声明式事务管理，" +
        "全局异常处理器将业务异常统一捕获并转换为契约约定的错误响应。";

    List<StackItem> items = List.of(
        new StackItem("language", "Java", String.valueOf(Runtime.version().feature()),
            "/static/icons/java.svg"),
        new StackItem("framework", "Spring Boot", SpringBootVersion.getVersion(),
            "/static/icons/springboot.svg"),
        new StackItem("database", dbName, dbVer,
            "/static/icons/" + dbIcon + ".svg")
    );

    long uptime = ManagementFactory.getRuntimeMXBean().getUptime() / 1000;
    HealthData data = new HealthData("ok", props.getServiceName(), props.getServiceVersion(),
        uptime, new StackInfo("backend", summary, items));
    return Envelope.ok(data);
  }

  @GetMapping("/avatars")
  public Envelope<List<Avatar>> avatars() {
    return Envelope.ok(avatars.list());
  }
}

File: backends/nest/src/health/health.controller.ts
import { Controller, Get } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { AppConfig } from '../config/configuration';
import { loadAvatars } from './avatar.service';

const startedAt = Date.now();

@Controller('api/v1')
export class HealthController {
  constructor(private readonly config: ConfigService<AppConfig>) {}

  @Get('health')
  health() {
    const driver = this.config.get('dbDriver', { infer: true });
    const serviceName = this.config.get('serviceName', { infer: true });
    const version = this.config.get('serviceVersion', { infer: true });

    const dbName = driver === 'postgres' ? 'PostgreSQL' : 'SQLite';
    const dbVer = driver === 'postgres' ? '16' : '3';

    const summary =
      '基于 Node.js + NestJS + TypeScript 核心骨架，选用 TypeORM 作为数据库对象关系映射工具，' +
      'Passport 进行 JWT 身份验证，class-validator 驱动声明式数据校验，' +
      '同时支持 PostgreSQL 与 SQLite 双数据库驱动切换。' +
      '采用 TypeScript 强类型约束和面向对象元数据设计，' +
      '借助 NestJS 的依赖注入容器与模块系统，为 Node.js 环境提供企业级架构标准。' +
      'TypeORM 创新设计了跨库列类型辅助机制，在 PostgreSQL 和 SQLite 间自动适配类型，' +
      '服务启动过程只负责连接已准备好的数据库，schema 由仓库级 scripts/db 维护。' +
      '通过 DTO 属性装饰器完成请求边界声明，' +
      '全局 ValidationPipe 拦截非法请求并过滤多余字段，在请求抵达业务层前构筑类型安全边界。' +
      '高度遵循 NestJS 标准的 AOP 生命周期，' +
      'Passport JWT Strategy 配合守卫实现鉴权拦截、拦截器完成统一响应包装、' +
      '全局过滤器捕获异常并映射为契约约定的错误响应。';

    return {
      status: 'ok',
      service: serviceName,
      version,
      uptimeSeconds: Math.floor((Date.now() - startedAt) / 1000),
      stack: {
        kind: 'backend',
        summary,
        items: [
          { role: 'language', name: 'TypeScript', version: '5.x', iconUrl: '/static/icons/typescript.svg' },
          { role: 'framework', name: 'NestJS', version: '11.x', iconUrl: '/static/icons/nestjs.svg' },
          { role: 'database', name: dbName, version: dbVer, iconUrl: `/static/icons/${dbName.toLowerCase()}.svg` },
        ],
      },
    };
  }

  @Get('avatars')
  avatars() {
    return loadAvatars();
  }
}
